File: decorator/java/FileDataSource.java
package Decorator;

// Temel işlemlerin tanımlandığı IDataSource arayüzünü uygulayan sınıftır.
// UML diyagramındaki ConcreteComponent sınıfına denk gelmektedir.
class FileDataSource implements IDataSource {
    private String _fileName;

    public FileDataSource(String fileName) {
        _fileName = fileName;
    }

    public String getFileName() {
        return _fileName;
    }

    public void readData() {
        System.out.println(_fileName + " readed.");
    }

    public void writeData(Object data) {
        System.out.println("data was written to " + _fileName);
    }
}

File: decorator/typescript/file-data-source.ts
import {IDataSource} from "./data-source.interface";

// Temel işlemlerin tanımlandığı IDataSource arayüzünü uygulayan sınıftır.
// UML diyagramındaki ConcreteComponent sınıfına denk gelmektedir.
export class FileDataSource implements IDataSource {
  private readonly fileName: string;

  constructor(fileName: string) {
    this.fileName = fileName;
  }

  getFileName(): string {
    return this.fileName;
  }

  readData(): void {
    console.log(`${this.fileName} readed.`);
  }

  writeData(data: object): void {
    console.log(`Data was written to ${this.fileName}`);
  }
}
